mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java

```diff
@@ -523,6 +528,18 @@ private Date toISODate(String value) throws ParseException {
         return cal.getTime();
     }
 
+
+    private boolean matches(SearchQuery.AttachmentCriterion criterion, MailboxMessage message) throws UnsupportedSearchException {
+        boolean mailHasAttachments = message.getAttachments().stream().anyMatch(new Predicate<MessageAttachment>() {
```

Comment:
You can have a helper class in mailbox/store that does this.

However I think you should iterate on the all body InputStream using Mime4J (a bit like done in ES) as some implementation, like JPA, don't support attachment retrieval.